### PR TITLE
Adds teleporter implant

### PR DESCRIPTION
**Pull request by monster860**
Adds a teleporter implant item.

- Select a target to teleport to before implanting
- After implanting, click the ui icon at any time to teleport yourself to the tracking beacon.
- 30 second cooldown.
- No, you cannot select emagged targets.
- If you are incapacitated in any way, you can go fuck yourself.
- When teleporting, there is a sound and a 2 second delay. If you are incapacitated in any way after that 2 second delay, you can go fuck yourself.

If needed, I can lockbox it. Or anything. **ANYTHING TO QUIT MY DEPENDENCE ON THE HAND TELE!!!**

:cl: monster860
rscadd: Adds the teleporter implant.
/:cl:

### CODE REVIEW

**Comment by FlattestGuitar** [comment]
This should be a traitor item, not an RnD implant, in my opinion.

**Comment by monster860** [comment]
Keep in mind, you have to get someone to put this in you. This would be pretty much useless as a traitor item.

**Comment by FlattestGuitar** [comment]
If you get yourself a destination tagger and send beacons down disposals you can pretty much access any place you want without leaving any evidence, regardless of whether people know you have the implant or not. 

This seems to me like it's way too powerful to be obtainable normally. Maybe at least force more experimentor torment fist to make this harder to unlock.

**Comment by monster860** [comment]
You can still pretty much access any place you want without leaving any evidence if you simply hack into the teleporter. It's not like you can't open any door in 5 seconds. Just listen for the boots.

File: code/modules/research/designs/medical_designs.dm [hunk 1]
```diff
@@ -280,6 +280,17 @@
...
 
+/datum/design/cyberimp_teleport
+	name = "Teleporter implant"
+	desc = "This implant will teleport you to a selected beacon when triggered."
+	id = "ci-teleporter"
+	req_tech = list("materials" = 7, "programming" = 5, "biotech" = 6, "bluespace" = 5)
+	build_type = PROTOLATHE | MECHFAB
+	construction_time = 60
+	materials = list(MAT_METAL = 200, MAT_GLASS = 200, MAT_SILVER = 600, MAT_GOLD = 600, MAT_PLASMA = 1000, MAT_DIAMOND = 2000)
+	build_path = /obj/item/organ/internal/cyberimp/brain/anti_stun
```
Comment (Fox-McCloud) [inline code/modules/research/designs/medical_designs.dm]: teleporter implant, now with stun reduction.

**Comment by Fox-McCloud** [comment]
No way. This is OP as hell.

Even if lockboxed, it's essentially a get out of jail free card from *anything* that isn't an instant stun.

Someone starts hacking you? Teleport away. Someone starts shooting at you? Teleporte away. Someone storming your office? Insta-teleport to.

Captain/Head getting attacked? Teleport to instantly.

**Comment by theColdflame** [comment]
:-1:  This is exactly what science doesn't need more of right now.

There are also some pretty huge balance issues with this, and it sounds like something that's not particularly fun to play against (see also: jaunt).

**Comment by taukausanake** [comment]
Not even if you up the requirements? Yeah this is something pretty dangerous for anyone to have, antags included. If this was made then you'd have to make it as expensive as a phazon I want to say, and even then it's still OP

**Comment by monster860** [comment]
>Someone storming your office? Insta-teleport to.

>Captain/Head getting attacked? Teleport to instantly.

*cough cough* hand tele *cough cough*

> Someone starts hacking you? Teleport away. Someone starts shooting at you? Teleporte away. 

There' fixed.

**Comment by tigercat2000** [comment]
the hand tele isn't even slightly a fair comparison and you know it
half the time it doesn't work
there are only two on station and you can't make more
you can drop it or lose it

File: code/modules/surgery/organs/augments_internal.dm [hunk 2]
```diff
@@ -144,6 +144,47 @@
...
+		if(R.syndicate == 1)
+			continue
+		var/tmpname = T.loc.name
+		if(areaindex[tmpname])
+			tmpname = "[tmpname] ([++areaindex[tmpname]])"
+		else
+			areaindex[tmpname] = 1
+		L[tmpname] = R
+
+	var/desc = input("Please select a location to lock in.", "[src]") in L
```
Comment (tigercat2000) [inline code/modules/surgery/organs/augments_internal.dm]: add a cancel button here

**Comment by monster860** [comment]
First of all, if it doesn't work, you can try to use it again. After all, you bring the hand tele with you into the portal, right?

Anti-drop implant fixes the "drop it or lose it" part.

Keep in mind, hand tele has more options. Select one thing if someone's invading your office, select another if someone's attacking the shit out of the captain.

File: code/modules/surgery/organs/augments_internal.dm (hunk 2 above)
Comment (monster860) [inline code/modules/surgery/organs/augments_internal.dm]: How?

File: code/modules/surgery/organs/augments_internal.dm [hunk 3]
```diff
@@ -144,6 +144,47 @@
...
+
+/obj/item/organ/internal/cyberimp/brain/teleport/attack_self(mob/living/user as mob) // Blatantly copy-pasted from teleguns.
+	var/list/L = list()
+	var/list/areaindex = list()
+
+	for(var/obj/item/device/radio/beacon/R in beacons)
+		var/turf/T = get_turf(R)
+		if (!T)
+			continue
+		if((T.z in config.admin_levels) || T.z > 7)
```
Comment (tigercat2000) [inline code/modules/surgery/organs/augments_internal.dm]: > MAX_Z

File: code/modules/surgery/organs/augments_internal.dm (hunk 2 above)
Comment (tigercat2000) [inline code/modules/surgery/organs/augments_internal.dm]:
```
var/desc = input("Please select a location to lock in.", "[src]") as null|anything in L
if(!desc)
    return
teleportlocs or whatever here

**Comment by tigercat2000** [comment]
> Keep in mind, hand tele has more options. Select one thing if someone's invading your office, select another if someone's attacking the shit out of the captain.

how is this not the same
it makes a list of every beacon in the world same as the hand tele
infact it's even better than the hand tele, because the hand tele needs a teleporter active and pointing at something

**Comment by monster860** [comment]
>because the hand tele needs a teleporter active and pointing at something

Having your teleporter not active changes nothing. Having it pointing at the wrong thing adds maybe 3-4 seconds to how long it takes for you to robust your way to the problem.

File: code/modules/surgery/organs/augments_internal.dm (hunk 3 above)
Comment (monster860) [inline code/modules/surgery/organs/augments_internal.dm]: Wait, so you think people should be able to teleport to the gateway missions?

**Comment by tigercat2000** [comment]
let's just analyze how powerful this is
 - instantly move to a pre-set beacon every 30 seconds
 - no one can follow you or know where you went if you are far away from the beacon
 - can be used anywhere in the world
 - doesn't rely on a teleporter being aimed at the right spot
 - metabuddy can hold onto that there beacon for you to give you a mobile spawn point literally at all times
 - can't ever possibly be stolen, countered, or lost
 - never has the danger of putting you in space unless someone spaces your precious beacon

**Comment by monster860** [comment]
Allow me to introduce you to the hand tele. I keep it on me all the time, like I keep my ID card on me.

- Select any teleporter you want, and it can send you through that teleporter
- Never, in my experience, has there been even one instance where there isn't at least one target on the hand tele that takes you to a functional teleporter.
- It has a small chance of sticking you in space, but if it does, you can literally just use it again.
- Can be used anywhere
- Can be used in intervals of time less than 30 seconds.
- More options.
- If you use it right, security has a really hard time following you. My method is to repeatedly spam "None (Dangerous)" until it spaces me. Have fun following me now, sec!

> can't ever possibly be stolen, countered, or lost

Stunbatons, hybrid tasers, telescopic batons, handcuffs, or anything else capable of incapacitating the user can stop a player with this implant.

>metabuddy can hold onto that there beacon for you to give you a mobile spawn point literally at all times

That can be done with a hand tele.

**Comment by monster860** [comment]
And here is a 2 second delay for security to hit the guy with a baton. There's the sound, then 2 seconds later, the teleport.

**Comment by Necaladun** [comment]
This is another powerful toy that science doesn't need. There's already enough power creep for sci.

File: code/modules/surgery/organs/augments_internal.dm (hunk 3 above)
Comment (TheDZD) [inline code/modules/surgery/organs/augments_internal.dm]: `MAX_Z` is the number of non-away mission Z levels.

**Comment by Kevak** [comment]
To be totally honest, even playing as a carrier scientist, this is waaaaay OP. Might make for a nice taitortot item with tweaks.

**Comment by TheDZD** [comment]
Science really doesn't need yet another powerful implant to toy around with, it is literally the most fucked department balance-wise, and unless it's redesigned, adding more and more cool things for them to make is only going to make it worse.

**Comment by monster860** [comment]
Even if lockboxed to security?

**Comment by tigercat2000** [comment]
security doesn't need a teleport implant, dear god

**Comment by Kevak** [comment]
Even if lockboxed. Especially if lockboxed. Cause that means anyone with an emag or armory access on their ID can use it.

**Comment by monster860** [comment]
What if I make it leave behind a portal someone can follow?

**Comment by tigercat2000** [comment]
what if it doesn't exist at all because science doesn't need to be making better hand teleporters

**Comment by Kevak** [comment]
I think that tweak would make it work if it were made into a traitor item so security can actually catch them. I do not think it would work as an R&D item.

**Comment by Necaladun** [comment]
Even if lockboxed, with double the delays for everything, and a portal, I still don't like the idea of science basically replicating hand teleporters. There are only 2 of them. 

There's too much power creep in science as it is.

**Comment by TheDZD** [comment]
There are two hand teleporters that exist on station. You can print an infinite number of these. There are two hand teleporters that exist on the station.

**Comment by Kevak** [comment]
Maybe if it were like, idunno, a Bluespace 10 item it would work for R&D. As in so copiously hard, that it isn't gonna happen more then once every eight months.

**Comment by Regen1** [comment]
First off. No, this is way too OP

Secondly, No way, the hand teleporter is a very rare item and a traitor objective, there's two on the station and those are reserved for the CE/RD and the captain, there can never be more than two unless admins spawn more. There's a reason you can't construct them in RnD.

Secondly, the hand teleporter keeps a portal open for some time, sec can follow you through it.

Thirdly, this has no malfunction or haywire effects, nothing happens if it's EMP'd or the user is wearing any bluespace gear, all other teleporting devices will have a large chance of fucking you over if you're wearing a BoH or Bluespace belt.

This is also one of those "self serve" PR's.

>ANYTHING TO QUIT MY DEPENDENCE ON THE HAND TELE!!!

**Comment by VampyrBytes** [comment]
@Regen1 
I agree with all four of your three points :wink:

**Pull request closed without merging**